### PR TITLE
Changes how forensic machine reports works

### PR DESCRIPTION
**Pull request by Casper3667**
- Fingerprint & fiber reports now show the location they were gathered from. Particularly useful when combining them.
- Fingerprint cards & fiber bags can now be renamed. This does not affect the source location that is displayed on reports.
- Door names will now appear on the reports if the door had one.
- Generally changed how the reports appear.
- The numbering of the reports is now per type (fiber, fingerprint, GSR, DNA) rather than all of them combined per mahine.
- Gunshot Residue (GSR) now shows all residue on the given item/mob, rather than only the most recent.

image

### CODE REVIEW

**Comment by Casper3667** [comment]
!review

File: code/modules/detectivework/microscope/dnascanner.dm [hunk 1]
```diff
@@ -110,13 +110,15 @@
...
 			for(var/blood in bloodsamp.dna)
-				data += "<span class='notice'>Blood type: [bloodsamp.dna[blood]]<br>\nDNA: [blood]<br><br></span>"
+				if(bloodsamp.dna[blood])
+					data += "<b>Blood type:</b> [bloodsamp.dna[blood]]<br>"
+				data += "<b>DNA:</b> [blood]<br><br>"
 		else
 			data += "No DNA found.<br>"
-		info = "<b>[src] analysis report #[report_num]</b><br>"
-		info += "<b>Scanned item:</b><br>[bloodsamp.name]<br>[bloodsamp.desc]<br><br>" + data
+		info = "<b><font size=\"4\">[src] anaylsis report #[report_num]</font></b><HR>"
```
Comment (NonQueueingMatt) [inline code/modules/detectivework/microscope/dnascanner.dm:120]:
```suggestion
		info = "<b><font size=\"4\">[src] analysis report #[report_num]</font></b><HR>"
```

File: code/modules/detectivework/tools/sample_kits.dm [hunk 2]
```diff
@@ -46,8 +54,33 @@
 		if(merge_evidence(O, user))
 			qdel(O)
 		return 1
+	else if (O.ispen())
+		var/tmp_label = sanitizeSafe(input(user, "Enter a label for [name]", "Label", label_text), MAX_NAME_LEN)
+		if(length(tmp_label) > 20)
```
Comment (NonQueueingMatt) [inline code/modules/detectivework/tools/sample_kits.dm:59]: 20 should be MAX_NAME_LEN

File: code/modules/detectivework/tools/sample_kits.dm [hunk 3]
```diff
@@ -46,8 +54,33 @@
...
+			to_chat(user, SPAN_NOTICE("The label can be at most 20 characters long."))
```
Comment (NonQueueingMatt) [inline code/modules/detectivework/tools/sample_kits.dm:60]: ditto above

File: code/modules/mob/living/carbon/human/human_defines.dm [hunk 4]
```diff
@@ -95,7 +95,7 @@
 	var/hand_blood_color
 
 	var/list/flavor_texts = list()
-	var/gunshot_residue
+	var/gunshot_residue = list()
```
Comment (NonQueueingMatt) [inline code/modules/mob/living/carbon/human/human_defines.dm:98]:
```suggestion
	var/list/gunshot_residue = list()
```

**Review by NonQueueingMatt — APPROVED** [review]

File: code/modules/detectivework/microscope/dnascanner.dm [hunk 5]
```diff
@@ -110,13 +110,15 @@
 		//dna data itself
 		var/data = "No scan information available."
 		if(bloodsamp.dna != null)
-			data = "Spectometric analysis on provided sample has determined the presence of [bloodsamp.dna.len] strings of DNA.<br><br>"
+			data = "Spectometric analysis on provided sample has determined the presence of [bloodsamp.dna.len] string(s) of DNA.<br><br>"
```
Comment (Geevies) [inline code/modules/detectivework/microscope/dnascanner.dm:113]:
```suggestion
			data = "Spectometric analysis on provided sample has determined the presence of [length(bloodsamp.dna)] string\s of DNA.<br><br>"
```

File: code/modules/detectivework/microscope/microscope.dm [hunk 6]
```diff
@@ -43,47 +45,56 @@
...
-			pname = "GSR report #[report_num]"
-			info = "<b>Scanned item:</b><br>[swab.name]<br><br>"
-
-			if(swab.gsr)
-				info += "Residue from a [swab.gsr] bullet detected."
+			report_gsr_num++
+			pname = "GSR report #[report_gsr_num]"
+			info = "<b><font size=\"4\">GSR anaylsis report #[report_gsr_num]</font></b><HR>"
+			if(length(swab.gsr))
+				info += "Residue analysis of \the [swab.name] has determined the presence of the following residues.<ul>"
```
Comment (Geevies) [inline code/modules/detectivework/microscope/microscope.dm:57]:
```suggestion
				info += "Residue analysis of the [swab.name] has determined the presence of the following residues.<ul>"
```
\the doesn't work with obj.name

File: code/modules/detectivework/microscope/microscope.dm [hunk 7]
```diff
@@ -43,47 +45,56 @@
...
+				for(var/gsr in swab.gsr)
+					info += "<li>Residue from a [gsr] bullet detected."
```
Comment (Geevies) [inline code/modules/detectivework/microscope/microscope.dm:59]:
```suggestion
					info += "<li>Residue from a [gsr] bullet detected.</li>"
```

File: code/modules/detectivework/microscope/microscope.dm [hunk 8]
```diff
@@ -43,47 +45,56 @@
...
+			pname = "Fiber report #[report_fiber_num]"
 			if(fibers.evidence)
-				info = "Molecular analysis on [fibers.name] has determined the presence of unique fiber strings.<br><br>"
+				info = "<b><font size=\"4\">Fiber anaylsis report #[report_fiber_num]</font></b><HR>"
+				info += "<b>Source locations:</b> "
+				info += "[english_list(fibers.source, "no sources were found", ", ", ", ", "")].<br><br>"
+				info += "Molecular analysis on [fibers.name] has determined the presence of unique fiber strings.<ul>"
 				for(var/fiber in fibers.evidence)
-					info += SPAN_NOTICE("Most likely match for fibers: [fiber]<br><br>")
+					info += "<li><b>Most likely match for fibers:</b> [fiber]"
```
Comment (Geevies) [inline code/modules/detectivework/microscope/microscope.dm:74]:
```suggestion
					info += "<li><b>Most likely match for fibers:</b> [fiber]</li>"
```

File: code/modules/detectivework/microscope/microscope.dm [hunk 9]
```diff
@@ -43,47 +45,56 @@
...
-			pname = "Empty slide report #[report_num]"
+			pname = "Empty slide report #[report_fiber_num]"
 			info = "Evidence suggests that there's nothing in this slide."
 	else if(istype(sample, /obj/item/sample/print))
-		pname = "Fingerprint report #[report_num]"
-		info = "<b>Fingerprint analysis report #[report_num]</b>: [sample.name]<br>"
+		report_print_num++
+		pname = "Fingerprint report #[report_print_num]"
 		var/obj/item/sample/print/card = sample
+		info = "<b><font size=\"4\">Fingerprint anaylsis report #[report_print_num]</font></b><HR>"
```
Comment (Geevies) [inline code/modules/detectivework/microscope/microscope.dm:85]:
```suggestion
		info = "<b><font size=\"4\">Fingerprint analysis report #[report_print_num]</font></b><HR>"
```

File: code/modules/detectivework/microscope/microscope.dm [hunk 10]
```diff
@@ -43,47 +45,56 @@
...
-			info += "Surface analysis has determined unique fingerprint strings:<br><br>"
+			info += "Surface analysis has determined unique fingerprint strings:<ul>"
 			for(var/prints in card.evidence)
-				info += SPAN_NOTICE("Fingerprint string: ")
+				info += "<li><b>Fingerprint string: </b>"
 				if(!is_complete_print(prints))
 					info += "INCOMPLETE PRINT"
 				else
 					info += "[prints]"
 				info += "<br>"
```
Comment (Geevies) [inline code/modules/detectivework/microscope/microscope.dm:96]:
```suggestion
				info += "</li>"
```

File: code/modules/detectivework/tools/evidencebag.dm [hunk 11]
```diff
@@ -102,8 +102,8 @@
 /obj/item/evidencebag/attackby(obj/item/W as obj, mob/user as mob)
 	if(W.ispen() || istype(W, /obj/item/device/flashlight/pen))
 		var/tmp_label = sanitizeSafe(input(user, "Enter a label for [name]", "Label", label_text), MAX_NAME_LEN)
-		if(length(tmp_label) > 15)
-			to_chat(user, "<span class='notice'>The label can be at most 15 characters long.</span>")
+		if(length(tmp_label) > MAX_NAME_LEN)
+			to_chat(user, "<span class='notice'>The label can be at most [MAX_NAME_LEN] characters long.</span>")
```
Comment (Geevies) [inline code/modules/detectivework/tools/evidencebag.dm:106]:
```suggestion
			to_chat(user, SPAN_NOTICE("The label can be at most [MAX_NAME_LEN] characters long."))
```

File: code/modules/detectivework/tools/sample_kits.dm [hunk 12]
```diff
@@ -3,12 +3,18 @@
 	icon = 'icons/obj/forensics.dmi'
 	w_class = ITEMSIZE_TINY
 	var/list/evidence = list()
+	var/list/source = list()
```
Comment (Geevies) [inline code/modules/detectivework/tools/sample_kits.dm:6]: both of these can be lazylists

File: code/modules/detectivework/tools/sample_kits.dm [hunk 13]
```diff
@@ -46,8 +54,33 @@
 		if(merge_evidence(O, user))
 			qdel(O)
 		return 1
+	else if (O.ispen())
+		var/tmp_label = sanitizeSafe(input(user, "Enter a label for [name]", "Label", label_text), MAX_NAME_LEN)
+		if(length(tmp_label) > MAX_NAME_LEN)
+			to_chat(user, SPAN_NOTICE("The label can be at most [MAX_NAME_LEN] characters long."))
```
Comment (Geevies) [inline code/modules/detectivework/tools/sample_kits.dm:60]:
```suggestion
			to_chat(user, SPAN_WARNING("The label can be at most [MAX_NAME_LEN] characters long."))
```

File: code/modules/detectivework/tools/sample_kits.dm [hunk 14]
```diff
@@ -46,8 +54,33 @@
...
+		else
+			to_chat(user, SPAN_NOTICE("You set the label to \"[tmp_label]\"."))
+			label_text = tmp_label
+			update_name_label()
+		return
 	return ..()
 
+/obj/item/sample/proc/update_name_label()
+	if(label_text == "")
+		return
```
Comment (Geevies) [inline code/modules/detectivework/tools/sample_kits.dm:70]:
```suggestion
		name = initial(name)
```

File: code/modules/detectivework/tools/sample_kits.dm [hunk 15]
```diff
@@ -46,8 +54,33 @@
...
+	else
+		name = "[initial(name)] ([label_text])"
+
+/obj/item/sample/New(var/newloc, var/atom/supplied)
```
Comment (Geevies) [inline code/modules/detectivework/tools/sample_kits.dm:74]: use Initialize

File: code/modules/mob/living/carbon/human/human_defines.dm [hunk 16]
```diff
@@ -95,7 +95,7 @@
 	var/hand_blood_color
 
 	var/list/flavor_texts = list()
-	var/gunshot_residue
+	var/list/gunshot_residue = list()
```
Comment (Geevies) [inline code/modules/mob/living/carbon/human/human_defines.dm:98]: lazylist

File: code/modules/projectiles/guns/projectile.dm [hunk 17]
```diff
@@ -101,10 +101,12 @@
 		var/mob/living/carbon/human/H = loc
 		if(istype(H))
 			if(!istype(H.gloves, /obj/item/clothing))
-				H.gunshot_residue = chambered.caliber
+				if(!H.gunshot_residue) H.gunshot_residue = list()
+				H.gunshot_residue |= chambered.caliber
```
Comment (Geevies) [inline code/modules/projectiles/guns/projectile.dm:105]: there should be a LAZYDISTINCTADD you can replace both of these lines with

File: code/modules/projectiles/guns/projectile.dm [hunk 18]
```diff
@@ -101,10 +101,12 @@
...
 			else
 				var/obj/item/clothing/G = H.gloves
-				G.gunshot_residue = chambered.caliber
+				if(!G.gunshot_residue) G.gunshot_residue = list()
+				G.gunshot_residue |= chambered.caliber
```
Comment (Geevies) [inline code/modules/projectiles/guns/projectile.dm:109]: same as above

**Review by Geevies — CHANGES_REQUESTED** [review]

**Review by Geevies — APPROVED** [review]

**Pull request merged**